jgrapht-core/src/main/java/org/jgrapht/alg/connectivity/DulmageMendelsohnDecomposition.java

```diff
@@ -0,0 +1,303 @@
+/*
+ * (C) Copyright 2018-2018, by CAE Tech Limited and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.alg.connectivity;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import org.jgrapht.Graph;
+import org.jgrapht.alg.interfaces.MatchingAlgorithm.Matching;
+import org.jgrapht.alg.interfaces.StrongConnectivityAlgorithm;
+import org.jgrapht.alg.matching.HopcroftKarpMaximumCardinalityBipartiteMatching;
+import org.jgrapht.graph.AsSubgraph;
+import org.jgrapht.graph.DefaultDirectedGraph;
+import org.jgrapht.graph.DefaultEdge;
+import org.jgrapht.graph.EdgeReversedGraph;
+import org.jgrapht.graph.builder.GraphBuilder;
+import org.jgrapht.traverse.DepthFirstIterator;
+
+/**
+ * The Dulmage-Mendelsohn Decomposition
+ * (https://en.wikipedia.org/wiki/Dulmage%E2%80%93Mendelsohn_decomposition)
+ * partitions a bipartite graph into subsets. One subset is dominated by the
+ * first partition, one by the second partition, and the third is the remaining
+ * vertices.
+ *
+ * A fine decomposition can be performed that additionally performs a
+ * strongly-connected-components algorithm on the remaining subset.
+ * 
+ * The implementation is based on:
+ * Bunus P., Fritzson P., Methods for Structural Analysis and Debugging of Modelica Models, 2nd International Modelica Conference 2002
+ * It has a running time of $O(V + E)$.
+ * 
+ * @author Peter Harman
+ * @param <V> Vertex type
+ * @param <E> Edge type
+ */
+public class DulmageMendelsohnDecomposition<V, E> {
+
+    private final Graph<V, E> graph;
+    private final Set<V> partition1;
+    private final Set<V> partition2;
+
+    /**
+     * Construct the algorithm for a given (bipartite) graph and it's partitions
+     *
+     * @param graph bipartite graph
+     * @param partition1 the first partition of vertices in the bipartite graph
+     * @param partition2 the second partition of vertices in the bipartite graph
+     */
+    public DulmageMendelsohnDecomposition(Graph<V, E> graph, Set<V> partition1, Set<V> partition2) {
```

Comment:
@jkinable yes, it would be great if someone adopted #498 and completed it since it's very close to being done.  But let's keep that separate from this PR.